Disable chat for the first 5s (configurable) after joining a server.

pchote opened the pull request:
Implements the same feature as the serverside hotfix, but now with client-side integration to disable the chat box until the timeout has expired. The delay has been reduced to 5s and is now configurable by the server host.

pchote commented:
Summary from a Discord discussion with @dragunoff:
* It is important that we provide some UI feedback indicating why the chat input is disabled and when it will be restored
* The chat input should take focus when it is restored, provided nothing else is already focused
* I want to keep the client-side integration general enough to be useful with server-side patches/hotfixes (e.g. chat rate limiting, automatically disabling chat for blacklisted clients or people who are trying to trigger bugs).

I'll try to work the first two points in and also fix the style error.

pchote commented:
Updated. This now works the same way as the server hotfix, filtering individual messages instead of having enable/disable events. The client-side disabling remains purely cosmetic, and players will see the server-generated rejection message if they bypass it.

<img width="366" alt="Screenshot 2021-07-04 at 16 13 24" src="https://user-images.githubusercontent.com/167819/124392562-3dace400-dcee-11eb-9efe-fbb9fda22c59.png">

abcdefg30 commented on OpenRA.Mods.Common/Widgets/Logic/Ingame/IngameChatLogic.cs:
```diff
@@ -317,6 +320,27 @@ void AddChatLine(TextNotification chatLine, bool suppressSound)
 				Game.Sound.PlayNotification(modRules, null, "Sounds", chatLineSound, null);
 		}
 
+		public override void Tick()
+		{
+			var chatWasEnabled = chatEnabled;
+			chatEnabled = Game.RunTime >= orderManager.ChatDisabledUntil && orderManager.ChatDisabledUntil != uint.MaxValue;
```
Is it possible to only tick this when we're in the lobby chat? (Can we re-use `isMenuChat` from the ctor?)

pchote replied:
We need this ingame too (testcase: set a longer timeout and start a game before it expires).

abcdefg30 replied:
Hm, I'm not sure if that is actually a good thing, but I guess this is to cover a spectator entering just as the game is starting.

pchote replied:
Also future hotfixes or features that might want to turn off a client's chat for some other reason (e.g. automated flood limiting or server moderator muting an abusive user).

teinarss commented on OpenRA.Game/Network/OrderManager.cs:
```diff
@@ -49,6 +49,7 @@ public sealed class OrderManager : IDisposable
 		readonly List<TextNotification> notificationsCache = new List<TextNotification>();
 
 		public IReadOnlyList<TextNotification> NotificationsCache => notificationsCache;
+		public long ChatDisabledUntil { get; internal set; }
```
IMO this doesn't belong on the OrderManager.

pchote replied:
Neither, but IMO it does belong with the `NotificationsCache` (i.e. the chat history). Theres no rush to merge this PR, so we can wait until your PR that you mentioned splits that out.

pchote replied:
We'll need to come up with a solution to move this forward so we can get it in for the playtest.

teinarss commented on OpenRA.Game/Server/Connection.cs:
```diff
@@ -26,6 +27,7 @@ public class Connection : IDisposable
 		public readonly int PlayerIndex;
 		public readonly string AuthToken;
 		public readonly EndPoint EndPoint;
+		public readonly Stopwatch ConnectionTimer = Stopwatch.StartNew();
```
I'm not a big fan either on having this here too, it would be better to have something listening on ConnectionStateChanged or solve it in another way.

pchote replied:
This is the server's Connection, so we dont have `ConnectionStateChanged` etc!

teinarss replied:
Doh, then this makes more sense!

pchote commented:
Rebased and updated. I moved `ChatDisabledUntil` into `TextNotificationsManager` which also isn't ideal, but is the only other relevant place until the hardcoded chat plumbing (which forces this to be defined *somewhere* in Game) can be improved.

pchote commented:
Rebased.

abcdefg30 commented on OpenRA.Game/Settings.cs:
```diff
@@ -101,6 +101,9 @@ public class ServerSettings
 		[Desc("For dedicated servers only, treat maps that fail the lint checks as invalid.")]
 		public bool EnableLintChecks = true;
 
+		[Desc("Delay in seconds before newly joined players can send chat messages.")]
```
```suggestion
		[Desc("Delay in milliseconds before newly joined players can send chat messages.")]
```
?

abcdefg30 reviewed (commented):
Can you update the launch scripts for the servers with the new parameter, too, please?

pchote committed:
Disable chat for the first 5s (configurable) after joining a server.

pchote commented:
Done.

abcdefg30 commented:
Needs a rebase now.

dragunoff commented:
:+1: for the UI/UX. I didn't spot any issue with it.

teinarss commented:
Needs a rebase tho

pchote commented:
Rebased.

abcdefg30 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/ebe2b65f5513c77fe14e747520770df65afda1d8...623c6182b09a21857f779a29caf8e30a4c9c3d48)
